substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/CardTable.java

```diff
@@ -222,71 +196,60 @@ private static boolean isCleanEntryAtIndex(Pointer table, UnsignedWord index) {
         return isCleanEntry(readEntryAtIndex(table, index));
     }
 
-    /** Convert a byte offset to a table index. */
     private static UnsignedWord tableOffsetToIndex(UnsignedWord offset) {
-        return offset.unsignedDivide(ENTRY_BYTES);
+        return offset.unsignedDivide(ENTRY_SIZE_BYTES);
     }
 
-    /** Convert a table index to a byte offset. */
     private static UnsignedWord indexToTableOffset(UnsignedWord index) {
-        return index.multiply(ENTRY_BYTES);
+        return index.multiply(ENTRY_SIZE_BYTES);
     }
 
-    /** Turn a Pointer into the memory into a table index. */
     private static UnsignedWord memoryPointerToIndex(Pointer memoryStart, Pointer memoryLimit, Pointer memoryPointer) {
         assert memoryStart.belowOrEqual(memoryLimit) : "memoryStart.belowOrEqual(memoryLimit)";
         assert memoryStart.belowOrEqual(memoryPointer) : "memoryStart.belowOrEqual(memoryPointer)";
         assert memoryPointer.belowOrEqual(memoryLimit) : "memoryPointer.belowOrEqual(memoryLimit)";
-        final UnsignedWord offset = memoryPointer.subtract(memoryStart);
+        UnsignedWord offset = memoryPointer.subtract(memoryStart);
         return memoryOffsetToIndex(offset);
     }
 
-    /** Visit the cards in a table. */
     private static boolean visitCards(Pointer table, UnsignedWord indexLimit, CardTable.Visitor visitor) {
-        boolean result = true;
-        result &= visitor.prologue(table, indexLimit);
-        if (result) {
-            for (UnsignedWord index = WordFactory.unsigned(0); index.belowThan(indexLimit); index = index.add(1)) {
-                final int entry = readEntryAtIndex(table, index);
-                result &= visitor.visitEntry(table, index, entry);
-                if (!result) {
-                    break;
-                }
+        if (!visitor.prologue(table, indexLimit)) {
+            return false;
+        }
+        for (UnsignedWord index = WordFactory.unsigned(0); index.belowThan(indexLimit); index = index.add(1)) {
+            int entry = readEntryAtIndex(table, index);
+            if (!visitor.visitEntry(table, index, entry)) {
+                return false;
             }
         }
-        result &= visitor.epilogue(table, indexLimit);
-        return result;
+        return visitor.epilogue(table, indexLimit);
```

Comment:
We don't have many of these prologue/epilogue visitors any more, but I think how they worked was that once one of the methods returned `false`, the walk would be aborted and no further methods would be called. Therefore, the visitor can do any post-walk tasks right away. Otherwise, it would be necessary to propagate to `epilogue` that the walk has been aborted earlier. I think that the original code was actually buggy and assumed that `&=` has short-circuit behavior.

Reply:
(I got that impression in several other locations using `&=`, too)

Reply:
Then it is probably best to just fully remove the prologue & epilogue methods (`prologue` seems unused and the `epilogue` can be moved to `visitEntry`).